clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java

```diff
@@ -0,0 +1,148 @@
+        }
+        InitProducerIdRequestData data = new InitProducerIdRequestData()
+                // Because we never include a producer epoch or ID in this request, we expect that some errors
+                // (such as PRODUCER_FENCED) will never be returned in the corresponding broker response.
+                // If we ever modify this logic to include an epoch or producer ID, we will need to update the
+                // error handling logic for this handler to accommodate these new errors.
+                .setProducerEpoch(ProducerIdAndEpoch.NONE.epoch)
+                .setProducerId(ProducerIdAndEpoch.NONE.producerId)
+                .setTransactionalId(key.idValue)
+                // Set transaction timeout to 1 since it's only being initialized to fence out older producers with the same transactional ID,
+                // and shouldn't be used for any actual record writes
+                .setTransactionTimeoutMs(1);
+        return new InitProducerIdRequest.Builder(data);
+    }
+
+    @Override
+    public ApiResult<CoordinatorKey, ProducerIdAndEpoch> handleSingleResponse(
+            Node broker,
+            CoordinatorKey key,
+            AbstractResponse abstractResponse
+    ) {
+        InitProducerIdResponse response = (InitProducerIdResponse) abstractResponse;
+
+        Errors error = Errors.forCode(response.data().errorCode());
+        if (error != Errors.NONE) {
+            return handleError(key, error);
+        }
+
+        Map<CoordinatorKey, ProducerIdAndEpoch> completed = Collections.singletonMap(key, new ProducerIdAndEpoch(
+                response.data().producerId(),
+                response.data().producerEpoch()
+        ));
+
+        return new ApiResult<>(completed, Collections.emptyMap(), Collections.emptyList());
+    }
+
+    private ApiResult<CoordinatorKey, ProducerIdAndEpoch> handleError(CoordinatorKey transactionalIdKey, Errors error) {
+        switch (error) {
+            case CLUSTER_AUTHORIZATION_FAILED:
+                return ApiResult.failed(transactionalIdKey, new ClusterAuthorizationException(
+                        "InitProducerId request for transactionalId `" + transactionalIdKey.idValue + "` " +
+                                "failed due to cluster authorization failure"));
+
+            case TRANSACTIONAL_ID_AUTHORIZATION_FAILED:
+                return ApiResult.failed(transactionalIdKey, new TransactionalIdAuthorizationException(
+                        "InitProducerId request for transactionalId `" + transactionalIdKey.idValue + "` " +
+                                "failed due to transactional ID authorization failure"));
+
+            // We intentionally omit cases for PRODUCER_FENCED, TRANSACTIONAL_ID_NOT_FOUND, and INVALID_PRODUCER_EPOCH
+            // since those errors should never happen when our InitProducerIdRequest doesn't include a producer epoch or ID
+            // and should therefore fall under the "unexpected error" catch-all case below
```

Comment:
Wouldn't this make more sense just before the `default:` clause?

Reply:
Yeah, that's a good point. Will move.